File: app/src/main/java/com/tt/handsomeman/util/SharedPreferencesUtils.java
package com.tt.handsomeman.util;

import android.content.SharedPreferences;

import com.tt.handsomeman.HandymanApp;

import javax.inject.Inject;

import dagger.Module;

@Module
public class SharedPreferencesUtils {

    private SharedPreferences sharedPreferences;

    @Inject
    public SharedPreferencesUtils(SharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    public <T> void put(String key, T data) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        if (data instanceof String) {
            editor.putString(key, (String) data);
        } else if (data instanceof Boolean) {
            editor.putBoolean(key, (Boolean) data);
        } else if (data instanceof Float) {
            editor.putFloat(key, (Float) data);
        } else if (data instanceof Integer) {
            editor.putInt(key, (Integer) data);
        } else if (data instanceof Long) {
            editor.putLong(key, (Long) data);
        } else {
            editor.putString(key, HandymanApp.getInstance().getGSon().toJson(data));
        }
        editor.apply();
    }

    @SuppressWarnings("unchecked")
    public <T> T get(String key, Class<T> anonymousClass) {
        if (anonymousClass == String.class) {
            return (T) sharedPreferences.getString(key, "");
        } else if (anonymousClass == Boolean.class) {
            return (T) Boolean.valueOf(sharedPreferences.getBoolean(key, false));
        } else if (anonymousClass == Float.class) {
            return (T) Float.valueOf(sharedPreferences.getFloat(key, 0));
        } else if (anonymousClass == Integer.class) {
            return (T) Integer.valueOf(sharedPreferences.getInt(key, 0));
        } else if (anonymousClass == Long.class) {
            return (T) Long.valueOf(sharedPreferences.getLong(key, 0));
        } else {
            return HandymanApp.getInstance().getGSon().fromJson(sharedPreferences.getString(key, ""), anonymousClass);
        }
    }
}

File: app/src/main/java/com/tt/handsomeman/util/SharedPreferencesUtils.kt
package com.tt.handsomeman.util

import android.content.SharedPreferences

import com.tt.handsomeman.HandymanApp

import javax.inject.Inject

import dagger.Module

@Module
class SharedPreferencesUtils @Inject
constructor(private val sharedPreferences: SharedPreferences) {

    fun <T> put(key: String, data: T) {
        val editor = sharedPreferences.edit()
        if (data is String) {
            editor.putString(key, data as String)
        } else if (data is Boolean) {
            editor.putBoolean(key, data as Boolean)
        } else if (data is Float) {
            editor.putFloat(key, data as Float)
        } else if (data is Int) {
            editor.putInt(key, data as Int)
        } else if (data is Long) {
            editor.putLong(key, data as Long)
        } else {
            editor.putString(key, HandymanApp.instance!!.gSon!!.toJson(data))
        }
        editor.apply()
    }

    operator fun <T> get(key: String, anonymousClass: Class<T>): T {
        return if (anonymousClass == String::class.java) {
            sharedPreferences.getString(key, "") as T?
        } else if (anonymousClass == Boolean::class.java) {
            java.lang.Boolean.valueOf(sharedPreferences.getBoolean(key, false)) as T
        } else if (anonymousClass == Float::class.java) {
            java.lang.Float.valueOf(sharedPreferences.getFloat(key, 0f)) as T
        } else if (anonymousClass == Int::class.java) {
            Integer.valueOf(sharedPreferences.getInt(key, 0)) as T
        } else if (anonymousClass == Long::class.java) {
            java.lang.Long.valueOf(sharedPreferences.getLong(key, 0)) as T
        } else {
            HandymanApp.instance!!.gSon!!.fromJson(sharedPreferences.getString(key, ""), anonymousClass)
        }
    }
}
